hathor/indexes/rocksdb_tx_group_index.py

```diff
@@ -0,0 +1,141 @@
+# limitations under the License.
+
+from abc import abstractmethod
+from typing import TYPE_CHECKING, Iterable, Optional, Sized, Tuple, TypeVar
+
+from structlog import get_logger
+
+from hathor.indexes.rocksdb_utils import RocksDBIndexUtils
+from hathor.indexes.tx_group_index import TxGroupIndex
+from hathor.transaction import BaseTransaction
+
+if TYPE_CHECKING:  # pragma: no cover
+    import rocksdb
+
+logger = get_logger()
+
+KT = TypeVar('KT', bound=Sized)
+
+
+class RocksDBTxGroupIndex(TxGroupIndex[KT], RocksDBIndexUtils):
+    """ Index of inputs/outputs by address.
+
+    This index uses rocksdb and the following key format:
+
+        key = [address][tx.timestamp][tx.hash]
+              |--34b--||--4 bytes---||--32b--|
+
+    It works nicely because rocksdb uses a tree sorted by key under the hood.
+
+    The timestamp must be serialized in big-endian, so ts1 > ts2 implies that bytes(ts1) > bytes(ts2),
+    hence the transactions are sorted by timestamp.
+    """
+
+    _KEY_SIZE: int
+    _CF_NAME: bytes
+
+    def __init__(self, db: 'rocksdb.DB', cf_name: bytes) -> None:
+        self.log = logger.new()
+        RocksDBIndexUtils.__init__(self, db, cf_name)
+
+    def force_clear(self) -> None:
+        self.clear()
+
+    @abstractmethod
+    def _serialize_key(self, key: KT) -> bytes:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _deserialize_key(self, _bytes: bytes) -> KT:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _extract_keys(self, tx: BaseTransaction) -> Iterable[KT]:
+        raise NotImplementedError
+
+    def _to_rocksdb_key(self, key: KT, tx: Optional[BaseTransaction] = None) -> bytes:
+        import struct
+        rocksdb_key = self._serialize_key(key)
+        assert len(rocksdb_key) == self._KEY_SIZE
+        if tx:
+            assert tx.hash is not None
+            assert len(tx.hash) == 32
+            rocksdb_key += struct.pack('>I', tx.timestamp) + tx.hash
+            assert len(rocksdb_key) == self._KEY_SIZE + 4 + 32
+        return rocksdb_key
+
+    def _from_rocksdb_key(self, rocksdb_key: bytes) -> Tuple[KT, int, bytes]:
+        import struct
+        assert len(rocksdb_key) == self._KEY_SIZE + 4 + 32
+        key = self._deserialize_key(rocksdb_key[:self._KEY_SIZE])
+        timestamp: int
+        (timestamp,) = struct.unpack('>I', rocksdb_key[self._KEY_SIZE:self._KEY_SIZE + 4])
+        tx_hash = rocksdb_key[38:]
+        # Should we differentiate `_KEY_SIZE` and `_SERIALIZED_KEY_SIZE`?
+        # assert len(key) == self._KEY_SIZE
+        assert len(tx_hash) == 32
+        return key, timestamp, tx_hash
+
+    def add_tx(self, tx: BaseTransaction) -> None:
+        """ Add tx inputs and outputs to the wallet index (indexed by its addresses).
+        """
+        assert tx.hash is not None
+
+        for key in self._extract_keys(tx):
+            self.log.debug('put key', key=key)
+            self._db.put((self._cf, self._to_rocksdb_key(key, tx)), b'')
```

Comment:
Is it missing the `publish` method?

Reply:
Not here. It was missing calling `publish` on rocksdb_address_index.py. Just added it!